### PR TITLE
Refactor MicroXS class and usage in IndependentOperator

### LINKED ISSUE #2580: Design of the `MicroXS` and `IndependentOperator` classes
I'm currently working on refactoring the `MicroXS` and `IndependentOperator` classes to support a wider variety of use cases. Namely, the `MicroXS` class currently stores one-group microscopic cross sections that are to be paired with one-group fluxes. For our [FES shutdown dose rate project](https://github.com/openmc-dev/fes-sdr-project), we need the ability to do activation calculations using a multigroup flux spectrum / cross sections. Furthermore, the `MicroXS` and `IndependentOperator` class are not currently structured to support multiple materials very well (you can have multiple materials in `IndependentOperator` but only a single `MicroXS`).

As I'm envisioning a refactor of these two classes, one issue I'm having trouble resolving in my mind is how to handle multiple materials. For example, I can imagine something like the following
```Python
# This involves running an OpenMC model to get microscopic cross sections
# by (material, nuclide, reaction, energy group) and fluxes by (material, energy group)
micros, fluxes = get_micros_and_fluxes(model)

op = openmc.deplete.IndependentOperator(materials, micros, fluxes)
timesteps = [...]
source_rates = [...]
integrator = openmc.deplete.PredictorIntegrator(op, timesteps, source_rates=source_rates)
```

The main issue is: _how do we specify the source rates for multiple materials_? The options are:
1. The source rate is specified as the actual fixed source rate in [neutrons/sec]. In this case, it would make sense for us to pass `fluxes` to the independent operator corresponding to a unity source rate, and then all the reaction rates would be scaled accordingly.
2. The source rate is the one-group flux in [neutrons/cm²-s]. This implies that we would have to have a separate instance of the integrator and operator for each material, so the code above would become:
    ```Python
    timesteps = [...]
    for mat, micro_xs, flux in zip(materials, micros, fluxes):
        op = openmc.deplete.IndependentOperator(mat, micro_xs, flux)
        source_rate = flux.sum()
        integrator = openmc.deplete.PredictorIntegrator(op, timesteps, source_rates=source_rate)
    ```

Option 2 is arguably closer to how codes like FISPACT work. The multigroup flux provided to the operator has arbitrary units and everything is normalized to the total flux in [n/cm²-s]. Option 1 feels syntactically cleaner and more consistent with `CoupledOperator`, but it does imply that the units on `fluxes` matter.

Anyone have thoughts on this? @eepeterson @shimwell @yardasol @pshriwise

### PR DESCRIPTION
**Pull request by paulromano**
# Description

This PR refactors the `MicroXS` class and corresponding downstream usage in the `IndependentOperator` class. The full motivation is described in #2580 so I won't go into too much detail here other than to say that now:

- `MicroXS` no longer subclasses `pandas.DataFrame` and instead stores a numpy array as the `data` attribute
- `MicroXS` can now store multigroup cross sections
- A new `get_microxs_and_flux` function gives the user a way of getting both (multigroup) microscopic cross sections and corresponding fluxes for a given set of domains.
- The user can now explicitly specify which nuclides/reactions they want instead of always inferring this from the depletion chain
- The `IndependentOperator` class now requires that you pass it fluxes for each material to be depleted.

Altogether, these changes make it pretty trivial to switch from `CoupledOperator` to `IndependentOperator` — just call `get_microxs_and_flux` to get fluxes and microscopic cross sections and pass them to `IndependentOperator` with all the other usual info.

Closes #2580

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: openmc/deplete/microxs.py
Comment (shimwell) [inline openmc/deplete/microxs.py:94]:
```suggestion
    if energies is None:
        energy_filter = openmc.EnergyFilter([0.0, 100.0e6])
    elif isinstance(energies, str):
        energy_filter = openmc.EnergyFilter.from_group_structure(energies)
    else:
        energy_filter = openmc.EnergyFilter(energies)
```
I think making use of an ```elif``` here could remove one check in some case

File: openmc/deplete/independent_operator.py
Comment (yardasol) [inline openmc/deplete/independent_operator.py:348]: See general comments box for big spiel on this

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @paulromano,

Overall, I like these changes. See my in-line comments for suggestions.

I initially wrote the below text in an in-line comment for lines 345-348, but decided to put it in my general comments box so we could discuss it without it getting buried in the comment stack.

-------

It took me a while to convince myself that this was alright, but it seems that the units work out.

In the initial implementation, `_IndependentRateHelper._results_cache`  was filled by $\sigma^j_i * N_i$ for $j$ reactions and $i$ nuclides. We were able to do this by abusing the structure of the code in `OpenMCOperator` and `ChainFissionHelper`. and having the normalization factor output a flux for us in `n/cm^2-s` based on the equation:

$$
\phi = \frac{P}{\Sigma_i Q_i \sigma^f_i N_i}
$$

(https://docs.openmc.org/en/stable/usersguide/depletion.html#equation-fission-q), which would then be multiplied by the cross sections, yielding reaction rates. This allowed us to calculate reaction rates based on just the cross sections, the source rate (power), and the fission Q values.

Now because of the changes from previous PRs removing the need for `dilute_initial`, this process appears to have some extra steps, with multiplication and division of various factors related to converting barns to cm^2 and vice versa, so I think that the linked equation no longer holds true.

I'm a bit concerned about this, as it means that instead of calculating a flux based on the power, cross sections, fission Q values, and number of atoms, then multiplying that flux by the cross sections to get reaction rates, we are instead relying on MG flux values that we calculate at starting composition to get the reaction rates. These values stay the same no matter how the composition changes, which isn't really accurate.

I've been thinking of how to incorporate the MG cross sections into an equation similar to the one above. I've come up with the following:

$$
\phi_j = \frac{P}{\Sigma_i Q_{j,j} \sigma^f_{i,j} N_i}
$$

for energy group $j$.

We would calculate $\phi_j$ for all $j$ using `_normalization_helper.factor()`, then multiply each cross section by it's corresponding `\phi_j`, then sum over all `j` to get the reaction rates.

While the math does work out, we'd need to have energy-group-specific Q-values (which I'm not even sure exist, and if they do I imagine it would require using either the `openmc.data` module or NJOY), as well as flow control in `OpenMCOperator._calculate_reaction_rates()`, as `IndependentRateHelper.get_material_rates()` would return a quantity indexed by reaction, nuclide, AND energy group. We'd also need to modify the `ChainFissionHelper`.

What are your thoughts on this? Maybe I should open an issue for this?

**Comment by paulromano** [comment]
> I'm a bit concerned about this, as it means that instead of calculating a flux based on the power, cross sections, fission Q values, and number of atoms, then multiplying that flux by the cross sections to get reaction rates, we are instead relying on MG flux values that we calculate at starting composition to get the reaction rates. These MG flux values stay the same no matter how the composition changes, which isn't going to be accurate for all problems

If you are using a one-group flux and you have only a single material that is being depleted, the value of the user-provided flux is 100% inconsequential because everything gets normalized in order for the reaction rates to produce the correct amount of power. If you have a _multigroup_ flux, the flux will then determine the proper weighting between the provided multigroup cross sections. Similarly, if you have _multiple materials_ being depleted in `IndependentOperator`, the per-material fluxes provide proper weighting between the different materials. Without providing fluxes and using them when calculating the reaction rates, there's no way we can get the proper weighting. Note that in all cases the power _is_ still used to adjust the reaction rates at each step.

You are correct that perhaps our equations need updating -- I'll look at what we have written and think about that further.

> While the math does work out, we'd need to have flow control in `OpenMCOperator._calculate_reaction_rates()`, as `IndependentRateHelper.get_material_rates()` would return a quantity indexed by reaction, nuclide, AND energy group. We'd also need to modify the `ChainFissionHelper` to handle the additional dimension in the `fission_rates` argument.

I'm not sure I understand where you're going with this. Maybe we should hash it out over a call :smile:

File: openmc/deplete/microxs.py
Comment (paulromano) [reply openmc/deplete/microxs.py:222]: The default in numpy is that if you don't specify the trailing dimensions, they are assumed to be full, so this would really be equivalent to `self.data[i_nuc, i_rx, :]`

File: openmc/deplete/microxs.py
Comment (paulromano) [reply openmc/deplete/microxs.py:139]: This is a numpy-ism. When you have many dimensions, `...` can be used to mean "put `:` in all other dimensions", so in this case, any dimensions between the first and last.

File: openmc/deplete/microxs.py
Comment (paulromano) [reply openmc/deplete/microxs.py:33]: It is not part of PEP8 but it is something I am advocating for moving forward. I've kind of debated internally in my mind for a while whether it is really worth the trouble. What eventually convinced me that we ought to be adding more type hints is the ability of IDEs (e.g., VS Code) to give _much_ better information on types when those hints are present. This is a huge benefit for developers reading through code, and as you probably know most people spend more time reading code than writing code. How many times have you looked through the depletion module and wondered "what is the type of this variable"? Because I know for me personally that happens all the time. Of course, type hints can be enormously helpful for users too, particularly if they are writing a Python script with OpenMC using a good IDE.

**Comment by yardasol** [comment]
@paulromano 
> If you are using a one-group flux and you have only a single material that is being depleted, the value of the user-provided flux is 100% inconsequential because everything gets normalized in order for the reaction rates to produce the correct amount of power. If you have a _multigroup_ flux, the flux will then determine the proper weighting between the provided multigroup cross sections.

This makes sense, but wouldn't the weighting change over time as the neutron spectrum shifts?

**Comment by paulromano** [comment]
@yardasol Yes, if the material composition changes enough that it perturbs the neutron flux, the weight can change. That's one of the reasons we need to run transport solves at every step for fission reactor depletion. However, for fusion systems, in a lot of cases the material is getting irradiated and activated but doesn't change enough to actually perturb the flux (effectively, there's no "burnup"). In those cases the original weighting from time t=0 should hold just fine.

File: openmc/deplete/microxs.py
Comment (yardasol) [reply openmc/deplete/microxs.py:222]: Gotcha

**Comment by paulromano** [comment]
> 1\. does it make sense to "promote" `MicroXS` to outside the depletion module?

Yes, I think there's probably a good argument for this. I'd prefer to keep any sort of file restructuring for another PR just to limit the scope of each of them. Is that OK with you?

> 2\. Can we add a couple "TODO"s to the `MicroXS` class? One to be able to generate a `MicroXS` object directly from cross section libraries rather than running a transport simulation and another to be able to generate a burnup matrix directly from the  `MicroXS` object.

Already have plans for this :smile: But yes, I'll add some TODOs in this PR for good measure

> 3\. I foresee some memory problems for large R2S problems in the way `get_microxs_and_flux` works, so it might be worth thinking about how to tally the non-zero burnup matrix entries directly for every domain as well as the total flux in each domain.

You'll have to elaborate more, but I guess the only opportunity I see is if you were to do each nuclide separately since we know that some reactions won't apply to all nuclides. Is that what you were hinting at?

**Comment by eepeterson** [comment]
Sounds great to me - fine to leave the restructuring for another PR and glad to hear you're already way ahead of me on more features for `MicroXS`.

As for my point 3 re: memory issues: Yes, essentially doing each nuclide separately, but through some combination of preprocessing continuous energy cross sections into continuous energy transmutation fractions or doing it on the fly through something like a  "reaction sum" score since the transmutation pathways should be fewer than the total possible reaction pathways.

File: openmc/deplete/independent_operator.py
Comment (eepeterson) [inline openmc/deplete/independent_operator.py:50]: I think if `IndependentOperator` is going to still subclass `OpenMCOperator` then rather than `materials` being the first arg it should be `model` to parallel the `CoupledOperator` class. Then `fluxes` and  `micros` can be optional args that if not specified get computed by `get_microxs_and_flux`, then the optional kwarg `domains` could be added here with a sensible default that defaults to all cells or all materials in the model. What do you think?

Following on to this, I see there is a TODO in `model.py` to support `IndependentOperator`. Could we add that in here with a new kwarg to `Model.deplete` like `operator='coupled'` or `operator='independent'`? Then we could call `model.deplete(timesteps, operator='independent')` and get a cell-based or material-based depletion calculation that runs a single transport calculation if `micros` and `fluxes` aren't already computed somewhere else.

File: openmc/deplete/microxs.py
Comment (eepeterson) [inline openmc/deplete/microxs.py:27]: can we make this optional and default to all materials or all cells?

File: openmc/deplete/independent_operator.py
Comment (paulromano) [reply openmc/deplete/independent_operator.py:50]: I see your point about `OpenMCOperator` but the `IndependentOperator` class is really intended to be something that you can use "independent" of the OpenMC transport solver. `Model` is really a bit too specific to OpenMC as it contains geometry and other stuff that is not really relevant for `IndependentOperator`. That being said, I do like your idea about streamlining things with respect to having fluxes/micros auto-calculated when the user doesn't pass them. Again, I'd prefer leaving that for another PR just to not have too much scope creep here; we can create an issue for this after.

Also agreed that adding support for `IndependentOperator` in the `Model` class could be quite natural with these changes.

File: openmc/deplete/microxs.py
Comment (paulromano) [reply openmc/deplete/microxs.py:27]: The way it's currently designed, this might be a bit confusing because you get back lists of fluxes and micros corresponding to the order that was used in `domains`. If we default to something, the user won't have any idea of the order.

Comment (eepeterson) [reply openmc/deplete/microxs.py:27]: Should we instead return a `dict` with material, cell, or universe id as the key and a tuple of `(MicroXS, flux)` as the value? In all likelihood the order of `domains` means nothing in particular so I'd be in favor of making order not important at all.

Comment (shimwell) [reply openmc/deplete/microxs.py:27]: I guess knowing the cell id might be useful if we want to remove particular cell from the geometry for an R2S simulation with moving geometry.

File: docs/source/usersguide/depletion.rst
Comment (shimwell) [inline docs/source/usersguide/depletion.rst:216]: Do these fluxes need to be of a certain group structure e.g vitamin-j 175

File: openmc/deplete/microxs.py
Comment (paulromano) [reply openmc/deplete/microxs.py:27]: Sorry for the late follow up -- @eepeterson and I talked offline about this but just wanted to summarize the discussion. The reason I've chosen to return a list rather than a dict is mostly related to memory. If you have a long list of domains in a model that you need cross sections for, returning a dict would require more memory whereas a list keeps it as small as possible. If a user really wants a dictionary, it would easy enough for them to create one in a one-liner.

File: docs/source/usersguide/depletion.rst
Comment (paulromano) [reply docs/source/usersguide/depletion.rst:216]: They need to be in the same group structure as the microscopic cross sections since they get multiplied by one another:

https://github.com/paulromano/openmc/blob/a208df7633185c0cb14aa059a1c6b6bb1d5e0485/openmc/deplete/independent_operator.py#L348

**Comment by paulromano** [comment]
@yardasol @shimwell any further comments on this PR?

**Comment by shimwell** [comment]
No

> @yardasol @shimwell any further comments on this PR?

Looks great to me LGTM

**Comment by shimwell** [comment]
I think the conflict on this PR might be due to me replacing OrderedDicts with {}, sorry about that

**Comment by yardasol** [comment]
@paulromano have you come up with an equation to replace [equation (1)](https://docs.openmc.org/en/stable/usersguide/depletion.html#equation-fission-q) in our doc pages? It would be good if we could have this before we merge.

**Comment by yardasol** [comment]
@paulromano I like your update but we should perhaps state how the normalization factor is used to get reaction rates.

**Comment by yardasol** [comment]
@paulromano This is coming together. I think you are missing some terms in the new equation (specifically the `number / (1e24 * volume)` term that is called `atom_per_bcm`.

Also, I feel this may have been covered at this point in one of the previous PRs, but why are the units of flux in the `MicroXS` class `n-cm/src`? I recall there being a different term for this quantity but it is proportional to the flux by a constant factor.

File: docs/source/usersguide/depletion.rst
Comment (shimwell) [reply docs/source/usersguide/depletion.rst:216]: Seeing as there are a few other changes on the equations would it be ok to sneak one change in here

**Comment by paulromano** [comment]
> I think you are missing some terms in the new equation (specifically the `number / (1e24 * volume)` term that is called `atom_per_bcm`.

@yardasol The number density is already accounted for in the $N_{i,m}$ term (total number of atoms). When we tally "reaction rates" now, because we are using `multiply_density=False`, the tally values do not account for the number density of each nuclide (since it is $\phi\sigma$, not $\phi\Sigma$). Multiplying by `atom_per_bcm` in `OpenMCOperator._calculate_reaction_rates` gives us the reaction rate accounting for the density of each nuclide. Does that make sense?

**Comment by yardasol** [comment]
@eepeterson @paulromano thanks for those explanations. I was just asking because these units might make it more difficult for someone who wanted to manually provide those fluxes (i.e. from a separate code, which the original implementation of `MicroXS` supported),

File: docs/source/usersguide/depletion.rst
Comment (yardasol) [inline docs/source/usersguide/depletion.rst:209]: Maybe add a note specifying that the units of the flux are not the expected `[n/cm^2-s]`

**Review by yardasol — APPROVED** [review]
Good work @paulromano, just one final comment, but otherwise I think this is good to go. Looks like we are having some RTD issues, so should I hold off on merging?

File: docs/source/usersguide/depletion.rst
Comment (paulromano) [reply docs/source/usersguide/depletion.rst:209]: I'm hesitant to make a note about the units here -- if we were to change it in the class, it will be easy to forget to change it here too. The `IndependentOperator` class docstring is explicit on what units are expected.

**Comment by paulromano** [comment]
> Looks like we are having some RTD issues, so should I hold off on merging?

The RTD issue was fixed in #2667; I merged the develop branch into this one and it looks like RTD is happy now. @eepeterson @yardasol if you guys are happy, feel free to hit the big green button!

**Comment by eepeterson** [comment]
all checks passing and a couple approvals - seems good to me!

**Comment by Ebiwonjumi** [comment]
Hi @paulromano, concerning the get_microxs_flux function, it will be good to save the reaction rates and flux values from the temporary run to a statepoint file for restart purposes.

**Comment by eepeterson** [comment]
@Ebiwonjumi thanks for chiming in! This is an already merged PR so if there is additional functionality you'd like to see could you create a new issue if it isn't already there?

**Comment by Ebiwonjumi** [comment]
Got it. Thanks @eepeterson

**Pull request merged**
